### PR TITLE
Fix issue with updating maps defined as reference types

### LINKED ISSUE #39386: [Bug]: Issue with updating maps defined as reference types
### Description

$title.
```ballerina
import ballerina/io;

public type AtomTableSexpr map<string>;

AtomTableSexpr atoms = {};


public function main()  {
    foreach var [name, atom] in [["float", "String"]] {
        atoms[name] = atom;
    }
    io:println(atoms); //  print {}

}
```

### Steps to Reproduce

_No response_

### Affected Version(s)

_No response_

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by chiranSachintha**
Fixes #39386 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [x] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Review by warunalakshitha — COMMENTED** [review]
LGTM

**Review by MaryamZi — APPROVED** [review]

File: tests/jballerina-unit-test/src/test/resources/test-src/types/map/map-value.bal [hunk 1]
```diff
@@ -1,3 +1,15 @@
+public type Animal Mat;
+
+public type Mat map<string>;
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/map/map-value.bal:3]:
```suggestion
public type Animals Map;

public type Map map<string>;
```

File: tests/jballerina-unit-test/src/test/resources/test-src/types/map/map-value.bal [hunk 2]
```diff
@@ -1,3 +1,15 @@
...
+
+public type Map map<string>;
+
+function testUpdateMapValue() {
+    Animals animals = {};
+    foreach var [name, age] in [["animal1", "Cat"], ["animal2", "Dog"]] {
+        animals[name] = age;
+    }
+    assertEquality({"animal1":"Cat", "animal2":"Dog"}, animals);
+}
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/map/map-value.bal:11]: Why are we adding at the top?

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/MapUtils.java [hunk 3]
```diff
@@ -156,4 +143,24 @@ public static void checkIsMapOnlyOperation(Type mapType, String op) {
...
+                                       Object value) {
+        switch (mapType.getTag()) {
+            case TypeTags.MAP_TAG:
+                handleInherentTypeViolatingMapUpdate(value, (BMapType) mapType);
+                mapValue.put(fieldName, value);
+                break;
+            case TypeTags.RECORD_TYPE_TAG:
+                if (handleInherentTypeViolatingRecordUpdate(mapValue, fieldName, value, (BRecordType) mapType, false)) {
+                    mapValue.put(fieldName, value);
+                } else {
```
[the remark this answers is not in the document]
Comment (chiranSachintha) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/MapUtils.java]: Updated.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/MapUtils.java [hunk 4]
```diff
@@ -156,4 +143,24 @@ public static void checkIsMapOnlyOperation(Type mapType, String op) {
...
         }
     }
+
+    private static void updateMapValue(Type mapType, MapValue<BString, Object> mapValue, BString fieldName,
+                                       Object value) {
+        switch (mapType.getTag()) {
+            case TypeTags.MAP_TAG:
+                handleInherentTypeViolatingMapUpdate(value, (BMapType) mapType);
+                mapValue.put(fieldName, value);
+                break;
```
[the remark this answers is not in the document]
Comment (chiranSachintha) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/MapUtils.java]: Updated.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/MapUtils.java [hunk 5]
```diff
@@ -156,4 +143,23 @@ public static void checkIsMapOnlyOperation(Type mapType, String op) {
...
+                return;
+            case TypeTags.RECORD_TYPE_TAG:
+                if (handleInherentTypeViolatingRecordUpdate(mapValue, fieldName, value, (BRecordType) mapType, false)) {
+                    mapValue.put(fieldName, value);
+                    return;
+                }
+                mapValue.remove(fieldName);
+                return;
+            case TypeTags.TYPE_REFERENCED_TYPE_TAG:
+                updateMapValue(((BTypeReferenceType) mapType).getReferredType(), mapValue, fieldName, value);
```
Comment (MaryamZi) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/MapUtils.java:162]: Shall we add another test where the referred type is another type reference type?

```ballerina
public type StringMap map<string>;
public type Map StringMap;

Map x = {};

function testFn()  {
    x["greeting"] = "hello";
    // assertion
}
```

File: tests/jballerina-unit-test/src/test/resources/test-src/types/map/map-value.bal (hunk 2 above)
Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/map/map-value.bal:11]: There is no special reason, due to the adding of type definitions, I have added that at the top.

Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/map/map-value.bal:11]: We can add it at the bottom, will maintain some level of increasing complexity.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/MapUtils.java (hunk 5 above)
Comment (chiranSachintha) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/MapUtils.java:162]: This scenario covered in the example that was added. Right?

File: tests/jballerina-unit-test/src/test/resources/test-src/types/map/map-value.bal (hunk 2 above)
Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/map/map-value.bal]: Updated.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/MapUtils.java [hunk 6]
```diff
@@ -49,20 +49,7 @@ public class MapUtils {
...
-            case TypeTags.RECORD_TYPE_TAG:
-                if (handleInherentTypeViolatingRecordUpdate(mapValue, fieldName, value, (BRecordType) mapType,
-                        false)) {
-                    mapValue.put(fieldName, value);
-                } else {
-                    mapValue.remove(fieldName);
-                }
-                break;
-        }
+        updateMapValue(mapType, mapValue, fieldName, value);
```
Comment (HindujaB) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/MapUtils.java:52]:
```suggestion
        updateMapValue(mapValue.getType(), mapValue, fieldName, value);
```

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/MapUtils.java (hunk 5 above)
Comment (MaryamZi) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/MapUtils.java:162]: Ah yes, my bad. Shall we add one for just Map though?

```ballerina
function testFn()  {
    Map x = {};
    x["greeting"] = "hello";
    // assertion
}
```

File: tests/jballerina-unit-test/src/test/resources/test-src/types/map/map-value.bal [hunk 7]
```diff
@@ -1,3 +1,7 @@
+public type Animals Map;
+
+public type Map map<string>;
+
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/map/map-value.bal:4]: Types can also be immediately before the test.

Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/map/map-value.bal]: Updated

**Review by MaryamZi — APPROVED** [review]

File: tests/jballerina-unit-test/src/test/resources/test-src/types/map/map-value.bal [hunk 8]
```diff
@@ -73,3 +73,26 @@ function testEchoAny(any value)returns (string) {
     string stringVal = <string> value;
     return stringVal;
 }
+
+public type Animals Map;
+
+public type Map map<string>;
+
+function testUpdateMapValue() {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/map/map-value.bal:81]: Let's rename this to indicate type reference types.

**Pull request merged**